### ISSUE TITLE
Pipeline improvements

### ISSUE THREAD

**Issue by madonuko** [terrapkg/packages#8]
# TL;DR
- [x] Support multiple commits from the same push event
- [x] Run pipeline for commits from bots/pipeline
- [x] Disable building for unsupported arch
- [x] Can we generate `chkupdate.py` automatically? Or even maybe we can have anda bump the versions automatically?
- [ ] Better system for running builds again? (happens if you have pkgs with deps on other pkgs in terra)

---

# Multiple commits
## Bug
[This pipeline](https://github.com/terrapkg/packages/actions/runs/3274094272) from fce1fd2 did not build `switchboard-plug-onlineaccounts` (changed at parent 83fdfb2) but built `prismlauncher` (changed at parent 7d91d4e)

# Pipeline commits
## Bug
31c1dc6 committed by workflow [*Automatically check for updates*](https://github.com/terrapkg/packages/blob/03b3687c2b9cbd21e8af37acbac99c5f6030cb89/.github/workflows/update.yml) did not trigger github actions.
## Possible Fix
For `autobuild.yml`:
```yml
on:
  workflow_run:
    workflows: [Automatically check for updates]
    types:
      - completed
```

# Disable building for unsupported arch
## Solution
Maybe a file called `ignore_arch` listing out the arch to ignore? When `build` runs, we can just first check if arch is in the file

# Build reruns
## Idea
- Manual trigger: [*Build Packages*](../blob/main/.github/workflows/build.yml) pipeline (current solution)
  - Kinda suck when you mess up the params
- Comment on a commit?
  - Possibly get abused
- Push a tag (sth like `build.pkg-name-here`)
  - idk this is a bit dumb (but works well imo)

**Comment by lleyton** [comment]
>  Can we generate chkupdate.py automatically? Or even maybe we can have anda bump the versions automatically?
For this, I think there should be an Anda config option to set a source as the latest tag, or at least a separate Terra specific config.

**Comment by windowsboy111** [comment]
Just fixed the workflows so that it can run builds on commits by the automatic update system
Though I'm very aware of the horrible changelogs it's making

**Comment by getchoo** [comment]
> Can we generate chkupdate.py automatically? Or even maybe we can have anda bump the versions automatically?

it would probably be better for this to be in anda.hcl (if anda can support something like that), with fields for the constants that we already use (NAME, REPO, REGEX, etc). alternatively, maybe a json file with those fields that could be read into a single python script that would replace update.sh, which could in turn be easily generated with another python script.

(rough) example of an update.sh replacement:
```python
REQUIRED_FIELDS = ["name","repo","spec","link","api_info","update_type"]


class updateInfo:
	def __init__(self, file):
		try:
			with open(file) as f:
				self.data = json.load(f)
				self.name = data["name"]
				self.repo = data["repo"]
				self.regex = data["regex"] # can be for a commit or version tag
				self.spec = data["spec"]
				self.link = data["link"]
				self.api_info = data["api_info"] # what key is grabbed from json
				self.update_type = data["update_type"] # can be regular or nightly
				self.time = data["time"] # what times a nightly package should be built
		except OSError:
			raise OSError(f"unable to read {file}!")
		except json.JSONDecodeError:
			raise json.JSONDecodeError(f"unable to read json from {file}!")

	def check(self, val: str):
		return val in self.data
	
	def is_nightly(self):
		return "nightly" == self.update_type


queue = []
for dirpath, dirnames, filenames in os.walk(pathlib.Path.resolve()):
	if "update.json" in filenames:
		queue.append(os.path.join(dirpath, filenames))

if not len(queue): exit("no update information found")

for file in queue:
    info = None
    try:
        info = updateInfo(file)
    except (OSError, json.JSONDecodeError):
        continue

	# check for fields
	for val in REQUIRED_FIELDS:
		if not info.check(val):
			print(f"{val} field missing in {file}!")
			continue

		# check build type
		if info.is_nightly():
			if info.check("time"):
				if datetime.utcnow().hour not in info.time:
					print(f"{info.name}: skipping update check")
					continue
			else:
				print(f"{info.name}: No build times provided!")

		ver = requests.get(info.link).json()[info.api_info]
		f = open(info.spec, 'r')

		matches = re.findall(info.regex, f.read())
		if not len(matches):
			print(f"{info.name}: Failed to match regex!")
			continue
		cur = matches[0]

		if ver == cur:
			print(f"{info.name}: Up to date!")
			continue
		print(f"{info.name}: {cur} -> {ver}")

		# substitute commit if nightly
		if info.is_nightly():
			newspec = re.sub(info.regex, f'%global commit {ver}', f.read())
			f.close()
			f = open(info.spec, 'w')
			f.write(newspec)
		# use rpmdev-bumpspec if regular
		else:
			newspec = os.system(f'rpmdev-bumpspec -n {ver} {SPEC}')

		f.close()
```
 
 and for an update.json:
 ```json
 {
	"name": "foo",
	"repo": "https://github.com/foobar/foo.git",
	"regex": "r'Version:\s*([\.\d]+)\n'",
	"spec": "foo.spec",
	"link": "https://api.github.com/repos/foobar/foo/releases/latest",
	"api_info": "tag_name",
	"update_type": "regular"
}
```
